pkg/standalone-utils/test/LidoRelayer.test.ts

```diff
@@ -1,462 +1,950 @@
+        it('does not leave dust on the relayer', async () => {
+          const poolId = basePoolId;
+          const tokenIn = stETH;
+          const tokenOut = wstETH;
+          const amount = fp(1);
+
+          await relayer.connect(sender).multicall([
+            encodeBatchSwap({
+              swaps: [{ poolId, tokenIn, tokenOut, amount }],
+              sender: sender,
+              recipient: relayer,
+              outputReferences: { wstETH: toChainedReference(0) },
+            }),
+            encodeUnwrap(relayer.address, recipient.address, toChainedReference(0)),
+          ]);
+
+          expect(await stETH.balanceOf(relayer)).to.be.eq(0);
+          expect(await wstETH.balanceOf(relayer)).to.be.eq(0);
         });
+      });
+    });
 
-        it('exits the pool', async () => {
-          const receipt = await relayer
-            .connect(sender)
-            .exitPool(basePoolId, sender.address, sender.address, exitRequest);
+    describe('joinPool', () => {
+      function encodeJoin(params: {
+        poolId: string;
+        sender: Account;
+        recipient: Account;
+        assets: TokenList;
+        maxAmountsIn: BigNumberish[];
+        userData: string;
+        outputReference?: BigNumberish;
+      }): string {
+        return relayerLibrary.interface.encodeFunctionData('joinPool', [
+          params.poolId,
+          0, // WeightedPool
+          TypesConverter.toAddress(params.sender),
+          TypesConverter.toAddress(params.recipient),
+          {
+            assets: params.assets.addresses,
+            maxAmountsIn: params.maxAmountsIn,
+            userData: params.userData,
+            fromInternalBalance: false,
+          },
+          0,
+          params.outputReference ?? 0,
+        ]);
+      }
+
+      context('when the relayer is authorized', () => {
+        it('joins the pool', async () => {
+          const amount = fp(1);
+
+          const receipt = await relayer.connect(sender).multicall([
+            encodeWrap(sender.address, relayer.address, amount, toChainedReference(0)),
+            encodeApprove(wstETH, MAX_UINT256),
+            encodeJoin({
+              poolId: basePoolId,
+              assets: tokens,
+              sender: relayer,
+              recipient: recipient,
+              maxAmountsIn: tokens.map(() => MAX_UINT256),
+              userData: WeightedPoolEncoder.joinExactTokensInForBPTOut(
+                tokens.map((token) => (token === wstETH ? toChainedReference(0) : 0)),
+                0
+              ),
+            }),
+          ]);
 
           expectEvent.inIndirectReceipt(await receipt.wait(), vault.instance.interface, 'PoolBalanceChanged', {
             poolId: basePoolId,
-            liquidityProvider: sender.address,
+            liquidityProvider: relayer.address,
           });
         });
 
-        it('does not send wstETH to the recipient', async () => {
+        it('does not take wstETH from the sender', async () => {
```

Comment:
```suggestion
        it('does not take wstETH from the user', async () => {
```

We have been using 'user' to refer to the `sender` account (since the relayer is sometimes a sender in other contexts). I'd be fine with renaming `sender` and `recipient` to e.g. `senderUser` or `recipientUser`, for clarity.